File: src/com/cattailsw/nanidroid/LayoutManager.java
package com.cattailsw.nanidroid;

import android.widget.FrameLayout;
import android.content.Context;
import android.view.Gravity;
import android.widget.FrameLayout.LayoutParams;
import android.util.Log;

public class LayoutManager {
    private static final String TAG = "LayoutManager";
    private SakuraView sv = null;
    private KeroView kv = null;
    private Balloon bSakura = null;
    private Balloon bKero = null;
    private FrameLayout fl = null;
    private Context mCtx = null;

    private static LayoutManager _self = null;

    private LayoutManager(Context ctx) {
	mCtx = ctx.getApplicationContext();
    }

    public static LayoutManager getInstance(Context ctx) {
	if (_self == null )
	    _self = new LayoutManager(ctx);

	return _self;
    }

    public void setViews(FrameLayout f, SakuraView s, KeroView k, Balloon bS, Balloon bK){
	fl = f;
	sv = s;
	kv = k;
	bSakura = bS;
	bKero = bK;
    }

    public void checkAndUpdateLayoutParam() {
	int layoutWidth = fl.getWidth();
	int layoutHeight = fl.getHeight();
	if ( layoutHeight <= 0 || layoutWidth <= 0 )
	    return;

	int sH = sv.currentSurface.origH;
	int sW = sv.currentSurface.origW;
	int kH = kv.currentSurface.origH;
	int kW = kv.currentSurface.origW;

	// need to compute the maximum allowed width and height for both kero and sakura
	// first compare sW + kW and lW
	float wScale = 1.0f;
	if ( sW + kW > layoutWidth ) {
	    wScale = ((float) layoutWidth / (float)(sW + kW) );
	}

	float hScale = 1.0f;
	int vH = Math.max(sH, kH);
	if ( vH > layoutHeight ) {
	    hScale = ((float) layoutHeight / (float)vH );
	}
	
	float scale = Math.min(wScale, hScale);
// 	Log.d(TAG, "wScale:hScale="+wScale+":"+hScale+", [lH:lW]=["+layoutHeight+":"+layoutWidth+"],[sh:sw]="
// 	      + sH + ":" + sW +"]");
	int scaledSakuraHeight = (int) ( sH * scale);
	FrameLayout.LayoutParams lpS = new FrameLayout.LayoutParams((int)(sW * scale),
								    scaledSakuraHeight,
								    Gravity.BOTTOM | Gravity.RIGHT);
	sv.setLayoutParams(lpS);
	int scaledKeroHeight = (int)(kH * scale);
	int scaledKeroWidth = (int) (kW * scale);
	FrameLayout.LayoutParams lpK = new FrameLayout.LayoutParams(scaledKeroWidth,
								    scaledKeroHeight,
								    Gravity.BOTTOM | Gravity.LEFT);
	kv.setLayoutParams(lpK);

	// next compute balloon sizes...
	if ( scaledKeroHeight * 2 < scaledSakuraHeight ) {
	    int kbH = (Math.max( scaledKeroHeight, scaledSakuraHeight - scaledKeroHeight) );

	    FrameLayout.LayoutParams lpBK = new FrameLayout.LayoutParams(scaledKeroWidth, kbH, 
									 Gravity.BOTTOM | Gravity.LEFT);
	    lpBK.bottomMargin = scaledKeroHeight;
	    bKero.setLayoutParams(lpBK);
	    int bH = layoutHeight - scaledSakuraHeight;
	    
	    FrameLayout.LayoutParams lpBS = new FrameLayout.LayoutParams(layoutWidth, bH, 
									 Gravity.TOP | Gravity.RIGHT);
	    bSakura.setLayoutParams(lpBS);

	}
	else {
	    int bH = layoutHeight - scaledSakuraHeight;
	    int bW = layoutWidth / 2;
	    FrameLayout.LayoutParams lpBS = new FrameLayout.LayoutParams(bW, bH, Gravity.TOP | Gravity.RIGHT);
	    bSakura.setLayoutParams(lpBS);
	    FrameLayout.LayoutParams lpBK = new FrameLayout.LayoutParams(bW, bH, Gravity.TOP | Gravity.LEFT);
	    bKero.setLayoutParams(lpBK);
	}

	fl.invalidate();
    }

}

File: app/src/main/java/com/cattailsw/nanidroid/LayoutManager.kt
package com.cattailsw.nanidroid

import android.content.Context
import android.view.Gravity
import android.widget.FrameLayout

class LayoutManager private constructor(context: Context) {
    private val mCtx: Context = context.applicationContext
    private var sv: SakuraView? = null
    private var kv: KeroView? = null
    private var bSakura: Balloon? = null
    private var bKero: Balloon? = null
    private var fl: FrameLayout? = null

    companion object {
        private var _self: LayoutManager? = null

        @JvmStatic
        fun getInstance(ctx: Context): LayoutManager {
            if (_self == null) {
                _self = LayoutManager(ctx)
            }
            return _self!!
        }
    }

    fun setViews(f: FrameLayout?, s: SakuraView?, k: KeroView?, bS: Balloon?, bK: Balloon?) {
        fl = f
        sv = s
        kv = k
        bSakura = bS
        bKero = bK
    }

    fun checkAndUpdateLayoutParam() {
        val layout = fl ?: return
        val sakuraView = sv ?: return
        val keroView = kv ?: return
        val sakuraBalloon = bSakura ?: return
        val keroBalloon = bKero ?: return

        val layoutWidth = layout.width
        val layoutHeight = layout.height
        if (layoutHeight <= 0 || layoutWidth <= 0) return

        val sH = sakuraView.currentSurface?.origH ?: 300
        val sW = sakuraView.currentSurface?.origW ?: 200
        val kH = keroView.currentSurface?.origH ?: 300
        val kW = keroView.currentSurface?.origW ?: 200

        var wScale = 1.0f
        if (sW + kW > layoutWidth) {
            wScale = layoutWidth.toFloat() / (sW + kW).toFloat()
        }

        var hScale = 1.0f
        val vH = Math.max(sH, kH)
        if (vH > layoutHeight) {
            hScale = layoutHeight.toFloat() / vH.toFloat()
        }

        val scale = Math.min(wScale, hScale)

        val scaledSakuraHeight = (sH * scale).toInt()
        val scaledSakuraWidth = (sW * scale).toInt()
        val lpS = FrameLayout.LayoutParams(
            scaledSakuraWidth,
            scaledSakuraHeight,
            Gravity.BOTTOM or Gravity.RIGHT
        )
        sakuraView.layoutParams = lpS

        val scaledKeroHeight = (kH * scale).toInt()
        val scaledKeroWidth = (kW * scale).toInt()
        val lpK = FrameLayout.LayoutParams(
            scaledKeroWidth,
            scaledKeroHeight,
            Gravity.BOTTOM or Gravity.LEFT
        )
        keroView.layoutParams = lpK

        if (scaledKeroHeight * 2 < scaledSakuraHeight) {
            val kbH = Math.max(scaledKeroHeight, scaledSakuraHeight - scaledKeroHeight)
            var kbW = scaledKeroWidth
            if (scaledKeroWidth < layoutWidth - scaledSakuraWidth) {
                kbW = layoutWidth - scaledSakuraWidth
            }
            val lpBK = FrameLayout.LayoutParams(kbW, kbH, Gravity.BOTTOM or Gravity.LEFT).apply {
                bottomMargin = scaledKeroHeight
            }
            keroBalloon.layoutParams = lpBK

            val bH = layoutHeight - scaledSakuraHeight
            val lpBS = FrameLayout.LayoutParams(layoutWidth, bH, Gravity.TOP or Gravity.RIGHT)
            sakuraBalloon.layoutParams = lpBS
        } else {
            val bH = layoutHeight - scaledSakuraHeight
            val bW = layoutWidth / 2
            val lpBS = FrameLayout.LayoutParams(bW, bH, Gravity.TOP or Gravity.RIGHT)
            sakuraBalloon.layoutParams = lpBS
            val lpBK = FrameLayout.LayoutParams(bW, bH, Gravity.TOP or Gravity.LEFT)
            keroBalloon.layoutParams = lpBK
        }

        layout.invalidate()
    }
}
